tests/unit/specification/parser_test.py

```diff
@@ -2421,3 +2421,97 @@ def test_parse_error_invalid_range_aspect() -> None:
+    )
+
+    p1_kind = model.ModularInteger("P1::Kind", expr.Pow(expr.Number(2), expr.Number(16)))
+
+    p1_packet_structure = [
+        model.Link(model.INITIAL, model.Field("Kind")),
+        model.Link(
+            model.Field("Kind"),
+            model.Field("Payload"),
+            size=expr.Sub(expr.Last("Message"), expr.Last("Kind")),
+        ),
+        model.Link(model.Field("Payload"), model.FINAL),
+    ]
+
+    p1_packet_types = {
+        Field("Kind"): p1_kind,
+        Field("Payload"): OPAQUE,
+    }
+
+    p1_packet = model.Message("P1::Frame", p1_packet_structure, p1_packet_types)
+
+    p2_length = model.ModularInteger("P2::Length", expr.Pow(expr.Number(2), expr.Number(16)))
+
+    p2_packet_structure = [
+        model.Link(model.INITIAL, model.Field("Length")),
+        model.Link(model.Field("Length"), model.Field("Protocol")),
+        model.Link(
+            model.Field("Protocol"),
+            model.Field("Payload"),
+            size=expr.Mul(expr.Number(8), expr.Variable("Length")),
+        ),
+        model.Link(model.Field("Payload"), model.FINAL),
+    ]
+
+    p2_packet_types = {
+        Field("Length"): p2_length,
+        Field("Protocol"): p4_protocol_number,
+        Field("Payload"): OPAQUE,
+    }
+
+    p2_packet = model.Message("P2::Packet", p2_packet_structure, p2_packet_types)
+
+    p3_packet_structure = [
+        model.Link(model.INITIAL, model.Field("F1")),
+        model.Link(model.Field("F1"), model.Field("F2")),
+        model.Link(model.Field("F2"), model.FINAL),
+    ]
+
+    p3_packet_types = {
+        Field("F1"): model.ModularInteger("P3::T", expr.Pow(expr.Number(2), expr.Number(32))),
+        Field("F2"): p2_length,
+    }
+
+    p3_packet = model.Message("P3::Packet", p3_packet_structure, p3_packet_types)
+
+    assert_messages_files([f"{SPEC_DIR}/in_p1.rflx"], [p2_packet, p3_packet, p1_packet])
+
+
+def test_parse_error_duplicate_spec_file_file() -> None:
+    p = parser.Parser()
+    p.parse(SPEC_DIR / "message_type.rflx")
+    with pytest.raises(
+        RecordFluxError,
+        match='parser: error: duplicate specification "tests/data/specs/subdir/message_type.rflx"\n'
+        'parser: error: previous specification "tests/data/specs/message_type.rflx"',
+    ):
+        p.parse(SPEC_DIR / "subdir/message_type.rflx")
+
+
+def test_parse_error_duplicate_spec_stdin_file() -> None:
+    p = parser.Parser()
+    p.parse_string(
+        """
+        package Message_Type is
+            type T is mod 2**32;
+            type M is
+                message
+                    F : T;
+                end message;
+        end Message_Type;
+        """
+    )
+    with pytest.raises(
+        RecordFluxError,
+        match='parser: error: duplicate specification "tests/data/specs/subdir/message_type.rflx"\n'
+        'parser: error: previous specification "<stdin>"',
+    ):
+        p.parse(SPEC_DIR / "subdir/message_type.rflx")
```

Comment:
Should conflicts on stdin be detected? The following code does not lead to an error:

```Python
    p = parser.Parser()
    p.parse_string(
        """
        package Message_Type is
            type T is mod 2**32;
            type M is
                message
                    F : T;
                end message;
        end Message_Type;
        """
    )
    p.parse_string(
        """
        package Message_Type is
            type T is mod 2**32;
        end Message_Type;
        """
    )
```

Reply:
I thought about this. The current implementation is solely based on file names (that's why `<stdin>` and `<stdin>` is not an error). For the case above to be detected, we'd need to compare the actual specification data whenever the file already exists in the specifications dict (which regularly happens when specs are included multiple times). Or we'd need to treat `stdin` specially and only compare the specs when the filename is `stdin`. I'm somewhat reluctant to introduce such a special handling only for the string parser (which we only use for tests).